File: java/patterns-use-cases/src/main/java/my/example/schedulingtasks/PaymentTracker.java
package my.example.schedulingtasks;

import static my.example.schedulingtasks.utils.Utils.escalateToHuman;
import static my.example.schedulingtasks.utils.Utils.sendReminderEmail;

import dev.restate.sdk.Restate;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;
import dev.restate.sdk.common.StateKey;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;
import java.time.Duration;
import my.example.schedulingtasks.utils.StripeEvent;

@VirtualObject
public class PaymentTracker {

  private static final StateKey<Boolean> PAID = StateKey.of("paid", Boolean.TYPE);
  private static final StateKey<Integer> REMINDER_COUNT =
      StateKey.of("reminder_count", Integer.TYPE);

  // Stripe sends us webhook events for invoice payment attempts
  @Handler
  public void onPaymentSuccess(StripeEvent event) {
    Restate.state().set(PAID, true);
  }

  @Handler
  public void onPaymentFailure(StripeEvent event) {
    var state = Restate.state();

    // Already paid, no need to send reminders
    if (state.get(PAID).orElse(false)) {
      return;
    }

    int remindersCount = state.get(REMINDER_COUNT).orElse(0);
    if (remindersCount < 3) {
      state.set(REMINDER_COUNT, remindersCount + 1);
      Restate.run("send reminder", () -> sendReminderEmail(event));

      // Schedule next reminder via a delayed self call
      Restate.virtualObjectHandle(PaymentTracker.class, Restate.key())
          .send(PaymentTracker::onPaymentFailure, event, Duration.ofDays(1));
    } else {
      // After three reminders, escalate to support team
      Restate.run("escalate", () -> escalateToHuman(event));
    }
  }

  public static void main(String[] args) {
    RestateHttpServer.listen(Endpoint.bind(new PaymentTracker()));
  }
}

File: kotlin/patterns-use-cases/src/main/kotlin/my/example/schedulingtasks/PaymentTracker.kt
package my.example.schedulingtasks

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.kotlin.endpoint.endpoint
import kotlin.time.Duration.Companion.days

@VirtualObject
class PaymentTracker {

  companion object {
    private val PAID = stateKey<Boolean>("paid")
    private val REMINDER_COUNT = stateKey<Int>("reminder_count")
  }

  // Stripe sends us webhook events for invoice payment attempts
  @Handler
  suspend fun onPaymentSuccess(event: StripeEvent) {
    state().set(PAID, true)
  }

  @Handler
  suspend fun onPaymentFailure(event: StripeEvent) {
    val state = state()

    // Already paid, no need to send reminders
    if (state.get(PAID) == true) {
      return
    }

    val remindersCount = state.get(REMINDER_COUNT) ?: 0
    if (remindersCount < 3) {
      state.set(REMINDER_COUNT, remindersCount + 1)
      runBlock { sendReminderEmail(event) }

      // Schedule next reminder via a delayed self call
      toVirtualObject<PaymentTracker>(objectKey()).request { onPaymentFailure(event) }.send(1.days)
    } else {
      // After three reminders, escalate to support team
      runBlock { escalateToHuman(event) }
    }
  }
}

fun main() {
  RestateHttpServer.listen(endpoint { bind(PaymentTracker()) })
}
